server/app/mutations/stacks/create.rb

```diff
@@ -44,7 +48,8 @@ def validate_services
     def execute
       attributes = self.inputs.clone
       grid = attributes.delete(:grid)
-      stack = Stack.create(name: self.name, grid: grid)
+      parent = attributes.delete(:parent_name)
+      stack = Stack.create(name: self.name, grid: grid, parent_name: parent)
```

Comment:
Childs are created before parents, that's why.

Reply:
Then it should validate that the `parent` stack does not exist.. :wink: 

```
$ kontena stack install test/deptest-nodeps.yaml
 [done] Creating stack deptest      
 [done] Triggering deployment of stack deptest     
 [done] Waiting for deployment to start     
 [done] Deploying service redis     
$ kontena stack install test/deptest.yaml 
> Installing dependency  as deptest-foo ..
> Redis version? : 3.2-alpine
 [done] Creating stack deptest-foo      
 [done] Triggering deployment of stack deptest-foo     
 [done] Waiting for deployment to start     
 [done] Deploying service db     
 [fail] Creating stack deptest      
 [error] 422 : Unprocessable Entity:
	name: deptest already exists
```

Although with the way that this loose stack parent-child linking works, that partially-installed `deptest-foo` stack actually ends up as a child of the originally installed `deptest` stack - the end result is very close to just `kontena stack upgrade deptest test/deptest.yaml`, just without touching the deptest stack and giving an error.

#### `deptest-nodeps.yaml`
```yaml
stack: test/deptest
services:
  redis:
    image: redis
```
#### `deptest.yaml
```yaml
stack: test/deptest
depends:
  foo:
    stack: kontena/redis
services:
  redis:
    image: redis
```
```

Reply:
And you can install any stack with `--parent-name foostack` to suddenly make it someones dependency. Actually I think you can do `kontena stack install --parent-name foostack --name foostack stack.yml` and crash everything.

Reply:
Validating that the `parent_name` stack does not exist and is the same as the `name` would prevent that.

ATM it looks like the server survives that kind of recursive stack, because it doesn't actually chase the parent/children stacks. The `Stack#parent_chain` method would get stuck in an infinite loop, but it isn't actually called anywhere outside of the specs (remove it?).

The `kontena stack ls` displays the tree wrong, and then `kontena stack upgrade` gets stuck in an infinite recursion... and so does `kontena stack rm`, so that kind of stack is tricky to remove :)

Reply:
BTW, by `validate` I mean something in the mutation `validate` method, like

```ruby
class Stacks::Create
  def validate
    if self.parent_name && self.grid.stacks.find_by(name: self.parent_name)
      add_error(:parent_name, :exists, "Parent stack #{self.parent_name} already exists")
    end
  end
end
```

The model validation is a good place for the `name == parent_name` check because that should never ever be true, but it's probably more convenient to do it in the mutation, and the resulting error message is more user-friendly.

Reply:
Okay, while the "validate not exists" would work for `kontena stack install`, it would break `kontena stack upgrade` - new child stacks can be created on upgrades, where the parent stack already exists:

```
$ kontena stack upgrade deptest test/deptest.yaml 
 [done] Parsing test/deptest.yaml     
 [done] Reading stack deptest from master     
Value for peers is not set. Substituting with an empty string.
> Installing new dependency /home/kontena/kontena/kontena/cli/test/deptest-child.yaml as deptest-bar ..
 [fail] Creating stack deptest-bar      
 [error] 422 : Unprocessable Entity:
	parent_name: Parent stack deptest already exists
```